### ISSUE TITLE
OctoPrint + Python 3.11

### ISSUE THREAD

**Issue by leigh-johnson** [bitsy-ai/printnanny-os#223]
## **Describe the bug**
Building OctoPrint from the `maintenance` branch to pull in support for Python 3.11. Seeing the following error:

```
Jan 26 08:43:40 office-v0-6 systemd[1]: Started OctoPrint.
Jan 26 08:43:42 office-v0-6 octoprint[815]: 2023-01-26 08:43:42,565 - octoprint.startup - CRITICAL - Could not initialize settings manager: No module named 'pydantic'
Jan 26 08:43:42 office-v0-6 octoprint[815]: 2023-01-26 08:43:42,568 - octoprint.startup - CRITICAL - There was a fatal error starting up OctoPrint.
Jan 26 08:43:42 office-v0-6 octoprint[815]: Could not initialize settings manager: No module named 'pydantic'
Jan 26 08:43:42 office-v0-6 octoprint[815]: There was a fatal error starting up OctoPrint.
Jan 26 08:43:42 office-v0-6 systemd[1]: octoprint.service: Deactivated successfully.
```

```
-----------------------
Build Info:  |
-----------------------
IMAGE_NAME = printnanny-release-image-raspberrypi4-64-20230126163001
DISTRO_PRETTY = PrintNanny Linux 0.6.1 (Feldspar)
VARIANT_NAME = PrintNanny OS Core
VARIANT_ID = core
DISTRO = printnanny
DISTRO_NAME = PrintNanny Linux
DISTRO_VERSION = 0.6.1
DISTRO_CODENAME = Feldspar
ID_LIKE = BitsyLinux
HOME_URL = https://printnanny.ai
BUG_REPORT_URL = https://github.com/bitsy-ai/printnanny-os/issues
YOCTO_VERSION = 4.1.2
YOCTO_CODENAME = Mickledore
MACHINE = raspberrypi4-64
TUNE_PKGARCH = cortexa72
-----------------------
Layer Revisions:      |
-----------------------
meta              = master:372960044f3c4180c767f686c316ee2a5b197865
meta-poky         = master:372960044f3c4180c767f686c316ee2a5b197865
meta-yocto-bsp    = master:372960044f3c4180c767f686c316ee2a5b197865
meta-raspberrypi  = mickledore-libcamera:b84ec14a8cece5a6a2b43ff8fc1a63f94560ec53
meta-oe           = master:ceceffcb1e3ef4f9ba7708f77c27e30a7aea61e5
meta-python       = master:ceceffcb1e3ef4f9ba7708f77c27e30a7aea61e5
meta-multimedia   = master:ceceffcb1e3ef4f9ba7708f77c27e30a7aea61e5
meta-bitsy        = mickledore-compat:44733bcf62329d1ce2f82a6ee6175b92f9fb7fba
meta-neural-network = mickledore:f19fa74c2f4c868eda1c48d709b66848f9cca5e8
meta-printnanny   = mickledore-compat:44733bcf62329d1ce2f82a6ee6175b92f9fb7fba
meta-networking   = master:ceceffcb1e3ef4f9ba7708f77c27e30a7aea61e5
meta-filesystems  = master:ceceffcb1e3ef4f9ba7708f77c27e30a7aea61e5
meta-initramfs    = master:ceceffcb1e3ef4f9ba7708f77c27e30a7aea61e5
meta-webserver    = master:ceceffcb1e3ef4f9ba7708f77c27e30a7aea61e5
meta-swupdate     = mickledore:c47a10be18412919f4c7041be05509ba97ac79f4
meta-microcontroller = mickledore:a848733698649628d0c3e34ae04ced0ee6ce3aee
```

**Comment by leigh-johnson** [comment]
Next up:
```
Jan 26 09:25:25 office-v0-6 octoprint[802]: 2023-01-26 09:25:25,423 - octoprint.startup - CRITICAL - Could not initialize settings manager: No module named 'class_doc'
Jan 26 09:25:25 office-v0-6 octoprint[802]: 2023-01-26 09:25:25,426 - octoprint.startup - CRITICAL - There was a fatal error starting up OctoPrint.
Jan 26 09:25:25 office-v0-6 octoprint[802]: Could not initialize settings manager: No module named 'class_doc'
```

**Comment by leigh-johnson** [comment]
:tada: Closing, fixes pushed to: https://github.com/bitsy-ai/meta-bitsy/pull/390

I'll merge this branch into the v0.6.x mainline after I've kicked all the tires.

**Comment by leigh-johnson** [comment]
Re-opening, the maintenance branch build is throwing an exception on startup. I'm guessing the move from `setuptools3_legacy` -> `setuptools` pep517 resulted in jinja files (and maybe all static assets) not being included in the binary wheel. 

```
Jan 26 13:48:11 office-v0-6 systemd[1]: Started OctoPrint.
Jan 26 13:48:16 office-v0-6 octoprint[817]: 2023-01-26 13:48:16,282 - py.warnings - WARNING - /usr/lib/python3.11/site-packages/requests/__init__.py:109: RequestsDependencyWarning: urllib3 (1.26.12) or chardet (No>
Jan 26 13:48:16 office-v0-6 octoprint[817]:   warnings.warn(
Jan 26 13:48:17 office-v0-6 octoprint[817]: 2023-01-26 13:48:17,516 - octoprint.startup - INFO - ******************************************************************************
Jan 26 13:48:17 office-v0-6 octoprint[817]: 2023-01-26 13:48:17,520 - octoprint.startup - INFO - Starting OctoPrint 1.8.6
Jan 26 13:48:17 office-v0-6 octoprint[817]: 2023-01-26 13:48:17,534 - octoprint.startup - INFO - ******************************************************************************
Jan 26 13:48:19 office-v0-6 octoprint[817]: 2023-01-26 13:48:19,152 - octoprint.util.connectivity.connectivity_checker - INFO - Connectivity state is currently: online
Jan 26 13:48:19 office-v0-6 octoprint[817]: 2023-01-26 13:48:19,153 - octoprint.util.connectivity.connectivity_checker - INFO - Connecting to 1.1.1.1:53 is working
Jan 26 13:48:19 office-v0-6 octoprint[817]: 2023-01-26 13:48:19,153 - octoprint.util.connectivity.connectivity_checker - INFO - Resolving octoprint.org is working
Jan 26 13:48:19 office-v0-6 octoprint[817]: 2023-01-26 13:48:19,188 - octoprint.plugin.core - INFO - Loading plugins from /usr/lib/python3.11/site-packages/octoprint/plugins, /home/printnanny/.octoprint/plugins an>
Jan 26 13:48:27 office-v0-6 octoprint[817]: 2023-01-26 13:48:27,058 - octoprint.plugin.core - INFO - Found 20 plugin(s) providing 20 mixin implementations, 40 hook handlers
Jan 26 13:48:27 office-v0-6 octoprint[817]: 2023-01-26 13:48:27,248 - octoprint.server.heartbeat - INFO - Starting server heartbeat, 900.0s interval
Jan 26 13:48:27 office-v0-6 octoprint[817]: 2023-01-26 13:48:27,559 - octoprint.server - INFO - Intermediary server started
Jan 26 13:48:27 office-v0-6 octoprint[817]: 2023-01-26 13:48:27,594 - octoprint.filemanager.storage - INFO - Initializing the file metadata for /home/printnanny/.octoprint/uploads...
Jan 26 13:48:27 office-v0-6 octoprint[817]: 2023-01-26 13:48:27,598 - octoprint.filemanager.storage - INFO - ... file metadata for /home/printnanny/.octoprint/uploads initialized successfully.
Jan 26 13:48:27 office-v0-6 octoprint[817]: 2023-01-26 13:48:27,615 - octoprint.server - INFO - Added new permission from plugin action_command_notification: PLUGIN_ACTION_COMMAND_NOTIFICATION_SHOW (needs: "Need(m>
Jan 26 13:48:27 office-v0-6 octoprint[817]: 2023-01-26 13:48:27,617 - octoprint.server - INFO - Added new permission from plugin action_command_notification: PLUGIN_ACTION_COMMAND_NOTIFICATION_CLEAR (needs: "Need(>
Jan 26 13:48:27 office-v0-6 octoprint[817]: 2023-01-26 13:48:27,627 - octoprint.server - INFO - Added new permission from plugin action_command_prompt: PLUGIN_ACTION_COMMAND_PROMPT_INTERACT (needs: "Need(method='r>
Jan 26 13:48:27 office-v0-6 octoprint[817]: 2023-01-26 13:48:27,628 - octoprint.server - INFO - Added new permission from plugin announcements: PLUGIN_ANNOUNCEMENTS_READ (needs: "Need(method='role', value='plugin_>
Jan 26 13:48:27 office-v0-6 octoprint[817]: 2023-01-26 13:48:27,629 - octoprint.server - INFO - Added new permission from plugin announcements: PLUGIN_ANNOUNCEMENTS_MANAGE (needs: "Need(method='role', value='plugi>
Jan 26 13:48:27 office-v0-6 octoprint[817]: 2023-01-26 13:48:27,630 - octoprint.server - INFO - Added new permission from plugin appkeys: PLUGIN_APPKEYS_ADMIN (needs: "Need(method='role', value='plugin_appkeys_adm>
Jan 26 13:48:27 office-v0-6 octoprint[817]: 2023-01-26 13:48:27,631 - octoprint.server - INFO - Added new permission from plugin appkeys: PLUGIN_APPKEYS_GRANT (needs: "Need(method='role', value='plugin_appkeys_use>
Jan 26 13:48:27 office-v0-6 octoprint[817]: 2023-01-26 13:48:27,634 - octoprint.server - INFO - Added new permission from plugin backup: PLUGIN_BACKUP_ACCESS (needs: "Need(method='role', value='plugin_backup_acces>
Jan 26 13:48:27 office-v0-6 octoprint[817]: 2023-01-26 13:48:27,642 - octoprint.server - INFO - Added new permission from plugin firmware_check: PLUGIN_FIRMWARE_CHECK_DISPLAY (needs: "Need(method='role', value='pl>
Jan 26 13:48:27 office-v0-6 octoprint[817]: 2023-01-26 13:48:27,654 - octoprint.server - INFO - Added new permission from plugin logging: PLUGIN_LOGGING_MANAGE (needs: "Need(method='role', value='plugin_logging_ma>
Jan 26 13:48:27 office-v0-6 octoprint[817]: 2023-01-26 13:48:27,664 - octoprint.server - INFO - Added new permission from plugin pi_support: PLUGIN_PI_SUPPORT_STATUS (needs: "Need(method='role', value='plugin_pi_s>
Jan 26 13:48:27 office-v0-6 octoprint[817]: 2023-01-26 13:48:27,684 - octoprint.server - INFO - Added new permission from plugin pluginmanager: PLUGIN_PLUGINMANAGER_LIST (needs: "Need(method='role', value='plugin_>
Jan 26 13:48:27 office-v0-6 octoprint[817]: 2023-01-26 13:48:27,694 - octoprint.server - INFO - Added new permission from plugin pluginmanager: PLUGIN_PLUGINMANAGER_MANAGE (needs: "Need(method='role', value='plugi>
Jan 26 13:48:27 office-v0-6 octoprint[817]: 2023-01-26 13:48:27,704 - octoprint.server - INFO - Added new permission from plugin pluginmanager: PLUGIN_PLUGINMANAGER_INSTALL (needs: "Need(method='role', value='plug>
Jan 26 13:48:27 office-v0-6 octoprint[817]: 2023-01-26 13:48:27,714 - octoprint.server - INFO - Added new permission from plugin softwareupdate: PLUGIN_SOFTWAREUPDATE_CHECK (needs: "Need(method='role', value='plug>
Jan 26 13:48:27 office-v0-6 octoprint[817]: 2023-01-26 13:48:27,723 - octoprint.server - INFO - Added new permission from plugin softwareupdate: PLUGIN_SOFTWAREUPDATE_UPDATE (needs: "Need(method='role', value='plu>
Jan 26 13:48:27 office-v0-6 octoprint[817]: 2023-01-26 13:48:27,731 - octoprint.server - INFO - Added new permission from plugin softwareupdate: PLUGIN_SOFTWAREUPDATE_CONFIGURE (needs: "Need(method='role', value='>
Jan 26 13:48:27 office-v0-6 octoprint[817]: 2023-01-26 13:48:27,890 - octoprint.plugins.softwareupdate - INFO - Loaded version cache from disk
Jan 26 13:48:30 office-v0-6 octoprint[817]: 2023-01-26 13:48:30,973 - octoprint.util.pip - INFO - Using "/home/printnanny/octoprint-venv/bin/python -m pip" as command to invoke pip
Jan 26 13:48:33 office-v0-6 octoprint[817]: 2023-01-26 13:48:33,962 - octoprint.util.pip - INFO - Version of pip is 22.3.1
Jan 26 13:48:34 office-v0-6 octoprint[817]: 2023-01-26 13:48:34,779 - octoprint.util.pip - INFO - pip installs to /home/printnanny/octoprint-venv/lib/python3.11/site-packages (writable -> no), --user flag needed ->
Jan 26 13:48:34 office-v0-6 octoprint[817]: 2023-01-26 13:48:34,780 - octoprint.util.pip - INFO - ==> pip ok -> NO!
Jan 26 13:48:34 office-v0-6 octoprint[817]: 2023-01-26 13:48:34,780 - octoprint.util.pip - ERROR - Cannot use this pip install, can't write to the install dir and also can't use --user for installing. Check your s>
Jan 26 13:48:34 office-v0-6 octoprint[817]: 2023-01-26 13:48:34,831 - octoprint.plugin.core - INFO - Initialized 20 plugin implementation(s)
Jan 26 13:48:34 office-v0-6 octoprint[817]: 2023-01-26 13:48:34,878 - octoprint.plugins.octoprint_nanny - INFO - Running on_settings_initialized handler
Jan 26 13:48:34 office-v0-6 octoprint[817]: 2023-01-26 13:48:34,906 - octoprint - INFO - Installing webcam compat overlay for configured default webcam <octoprint.webcams.ProvidedWebcam object at 0x7fa8c0d310>
Jan 26 13:48:34 office-v0-6 octoprint[817]: 2023-01-26 13:48:34,910 - octoprint.plugin.core - INFO - 20 plugin(s) registered with the system:
Jan 26 13:48:34 office-v0-6 octoprint[817]: |  Action Command Notification Support (bundled) = /usr/lib/python3.11/site-packages/octoprint/plugins/action_command_notification
Jan 26 13:48:34 office-v0-6 octoprint[817]: |  Action Command Prompt Support (bundled) = /usr/lib/python3.11/site-packages/octoprint/plugins/action_command_prompt
Jan 26 13:48:34 office-v0-6 octoprint[817]: |  Announcement Plugin (bundled) = /usr/lib/python3.11/site-packages/octoprint/plugins/announcements
Jan 26 13:48:34 office-v0-6 octoprint[817]: |  Anonymous Usage Tracking (bundled) = /usr/lib/python3.11/site-packages/octoprint/plugins/tracking
Jan 26 13:48:34 office-v0-6 octoprint[817]: |  Application Keys Plugin (bundled) = /usr/lib/python3.11/site-packages/octoprint/plugins/appkeys
Jan 26 13:48:34 office-v0-6 octoprint[817]: |  Backup & Restore (bundled) = /usr/lib/python3.11/site-packages/octoprint/plugins/backup
Jan 26 13:48:34 office-v0-6 octoprint[817]: |  Classic Webcam (bundled) = /usr/lib/python3.11/site-packages/octoprint/plugins/classicwebcam
Jan 26 13:48:34 office-v0-6 octoprint[817]: |  Core Wizard (bundled) = /usr/lib/python3.11/site-packages/octoprint/plugins/corewizard
Jan 26 13:48:34 office-v0-6 octoprint[817]: |  Discovery (bundled) = /usr/lib/python3.11/site-packages/octoprint/plugins/discovery
Jan 26 13:48:34 office-v0-6 octoprint[817]: |  Error Tracking (bundled) = /usr/lib/python3.11/site-packages/octoprint/plugins/errortracking
Jan 26 13:48:34 office-v0-6 octoprint[817]: |  Event Manager (bundled) = /usr/lib/python3.11/site-packages/octoprint/plugins/eventmanager
Jan 26 13:48:34 office-v0-6 octoprint[817]: |  File Check (2021.2.23) (bundled) = /usr/lib/python3.11/site-packages/octoprint_file_check
...skipping...
Jan 26 13:52:35 office-v0-6 octoprint[817]:     response = default_view()
Jan 26 13:52:35 office-v0-6 octoprint[817]:                ^^^^^^^^^^^^^^
Jan 26 13:52:35 office-v0-6 octoprint[817]:   File "/usr/lib/python3.11/site-packages/octoprint/server/views.py", line 661, in default_view
Jan 26 13:52:35 office-v0-6 octoprint[817]:     return preemptively_cached()
Jan 26 13:52:35 office-v0-6 octoprint[817]:            ^^^^^^^^^^^^^^^^^^^^^
Jan 26 13:52:35 office-v0-6 octoprint[817]:   File "/usr/lib/python3.11/site-packages/octoprint/server/util/flask.py", line 1199, in decorated_function
Jan 26 13:52:35 office-v0-6 octoprint[817]:     return f(*args, **kwargs)
Jan 26 13:52:35 office-v0-6 octoprint[817]:            ^^^^^^^^^^^^^^^^^^
Jan 26 13:52:35 office-v0-6 octoprint[817]:   File "/usr/lib/python3.11/site-packages/octoprint/server/util/flask.py", line 1281, in decorated_function
Jan 26 13:52:35 office-v0-6 octoprint[817]:     return f(*args, **kwargs)
Jan 26 13:52:35 office-v0-6 octoprint[817]:            ^^^^^^^^^^^^^^^^^^
Jan 26 13:52:35 office-v0-6 octoprint[817]:   File "/usr/lib/python3.11/site-packages/octoprint/server/util/flask.py", line 1291, in decorated_function
Jan 26 13:52:35 office-v0-6 octoprint[817]:     r = f(*args, **kwargs)
Jan 26 13:52:35 office-v0-6 octoprint[817]:         ^^^^^^^^^^^^^^^^^^
Jan 26 13:52:35 office-v0-6 octoprint[817]:   File "/usr/lib/python3.11/site-packages/octoprint/server/util/flask.py", line 937, in decorated_function
Jan 26 13:52:35 office-v0-6 octoprint[817]:     rv = f_with_duration(*args, **kwargs)
Jan 26 13:52:35 office-v0-6 octoprint[817]:          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Jan 26 13:52:35 office-v0-6 octoprint[817]:   File "/usr/lib/python3.11/site-packages/octoprint/server/util/flask.py", line 889, in f_with_duration
Jan 26 13:52:35 office-v0-6 octoprint[817]:     return f(*args, **kwargs)
Jan 26 13:52:35 office-v0-6 octoprint[817]:            ^^^^^^^^^^^^^^^^^^
Jan 26 13:52:35 office-v0-6 octoprint[817]:   File "/usr/lib/python3.11/site-packages/octoprint/server/util/flask.py", line 1210, in decorated_function
Jan 26 13:52:35 office-v0-6 octoprint[817]:     rv = f(*args, **kwargs)
Jan 26 13:52:35 office-v0-6 octoprint[817]:          ^^^^^^^^^^^^^^^^^^
Jan 26 13:52:35 office-v0-6 octoprint[817]:   File "/usr/lib/python3.11/site-packages/octoprint/server/util/flask.py", line 1235, in decorated_function
Jan 26 13:52:35 office-v0-6 octoprint[817]:     rv = f(*args, **kwargs)
Jan 26 13:52:35 office-v0-6 octoprint[817]:          ^^^^^^^^^^^^^^^^^^
Jan 26 13:52:35 office-v0-6 octoprint[817]:   File "/usr/lib/python3.11/site-packages/octoprint/server/views.py", line 651, in make_default_ui
Jan 26 13:52:35 office-v0-6 octoprint[817]:     r = make_response(render_template("index.jinja2", **render_kwargs))
Jan 26 13:52:35 office-v0-6 octoprint[817]:                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Jan 26 13:52:35 office-v0-6 octoprint[817]:   File "/usr/lib/python3.11/site-packages/flask/templating.py", line 147, in render_template
Jan 26 13:52:35 office-v0-6 octoprint[817]:     return _render(app, template, context)
Jan 26 13:52:35 office-v0-6 octoprint[817]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Jan 26 13:52:35 office-v0-6 octoprint[817]:   File "/usr/lib/python3.11/site-packages/flask/templating.py", line 130, in _render
Jan 26 13:52:35 office-v0-6 octoprint[817]:     rv = template.render(context)
Jan 26 13:52:35 office-v0-6 octoprint[817]:          ^^^^^^^^^^^^^^^^^^^^^^^^
Jan 26 13:52:35 office-v0-6 octoprint[817]:   File "/usr/lib/python3.11/site-packages/jinja2/environment.py", line 1301, in render
Jan 26 13:52:35 office-v0-6 octoprint[817]:     self.environment.handle_exception()
Jan 26 13:52:35 office-v0-6 octoprint[817]:   File "/usr/lib/python3.11/site-packages/jinja2/environment.py", line 936, in handle_exception
Jan 26 13:52:35 office-v0-6 octoprint[817]:     raise rewrite_traceback_stack(source=source)
Jan 26 13:52:35 office-v0-6 octoprint[817]:   File "/usr/lib/python3.11/site-packages/octoprint/templates/index.jinja2", line 172, in top-level template code
Jan 26 13:52:35 office-v0-6 octoprint[817]:     {% include 'dialogs/wizard.jinja2' %}
Jan 26 13:52:35 office-v0-6 octoprint[817]:     ^^^^^^^^^^^^^^^^^^^^^^^^^
Jan 26 13:52:35 office-v0-6 octoprint[817]:   File "/usr/lib/python3.11/site-packages/octoprint/templates/dialogs/wizard.jinja2", line 45, in top-level template code
Jan 26 13:52:35 office-v0-6 octoprint[817]:     {% include data.template ignore missing %}
Jan 26 13:52:35 office-v0-6 octoprint[817]:     ^^^^^^^^^^^^^^^^^^^^^^^^^
Jan 26 13:52:35 office-v0-6 octoprint[817]:   File "/usr/lib/python3.11/site-packages/octoprint/plugins/corewizard/templates/corewizard_webcam_wizard.jinja2", line 29, in top-level template code
Jan 26 13:52:35 office-v0-6 octoprint[817]:     {% include "snippets/settings/webcam/webcamStreamUrl.jinja2" %}
Jan 26 13:52:35 office-v0-6 octoprint[817]:     ^^^^^^^^^^^^^^^^^^^^^^^^^
Jan 26 13:52:35 office-v0-6 octoprint[817]:   File "/usr/lib/python3.11/site-packages/flask/templating.py", line 62, in get_source
Jan 26 13:52:35 office-v0-6 octoprint[817]:     return self._get_source_fast(environment, template)
Jan 26 13:52:35 office-v0-6 octoprint[817]:            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Jan 26 13:52:35 office-v0-6 octoprint[817]:   File "/usr/lib/python3.11/site-packages/flask/templating.py", line 98, in _get_source_fast
Jan 26 13:52:35 office-v0-6 octoprint[817]:     raise TemplateNotFound(template)
Jan 26 13:52:35 office-v0-6 octoprint[817]: jinja2.exceptions.TemplateNotFound: snippets/settings/webcam/webcamStreamUrl.jinja2
```

**Comment by leigh-johnson** [comment]
:checkered_flag: Verified OctoPrint installs cleanly in the latest nightly.

**Comment by leigh-johnson** [comment]
Re-opening, I'm seeing some reports of plugins failing to install on Python 3.11

**Comment by leigh-johnson** [comment]
A bit of context on why we ended up with Python 3.11 in the base install:

I needed to upgrade to a version of Rust that supports `AsyncInto` (1.64+). 1.66 was available in the mickledore branch of the yocto poky reference distribution, so I shifted the base from langdale -> mickledore. Python 3.10 was upgraded to Python 3.11 in mickledore as well.

I think the right move is to package a Python 3.10 recipe for mickledore. The GCC and Rust upgrades in mickledore have allowed me to move to the latest async-nats lib and catch a `sprintf` bug in Janus Gateway, plus the meta-raspberrypi layer is already constraining compat to mickledore. 

I'll tackle the Python 3.11 can of worms when the next poky LTS is cut in April-May 20223
